### DISCUSSION TITLE
Persistent storage

### DISCUSSION THREAD

**Discussion by defdefred** [VendelinSlezak/BleskOS#40] [Ideas]
Could be nice to have a persistent storage on the boot usb to keep some configuration between reboot and upgrade. Keyboard mapping for example, or vesa mode, or color of the UI, or maybe an alternate bitmap font.
What do you think?

**Comment by VendelinSlezak** [comment]
Yes, this is definitely good idea, but because now only thing that would be really useful to save is keyboard layout, so I will implement this when there will be more settings.

**Comment by defdefred** [comment]
2 partitions boot disk is already working.
[2partitionbootdisk.txt](https://github.com/VendelinSlezak/BleskOS/files/13814643/2partitionbootdisk.txt)

I need to try but dd-ing a new version to the boot disk without smashing the mbr should work don't you think?

**Comment by VendelinSlezak** [comment]
> I need to try but dd-ing a new version to the boot disk without smashing the mbr should work don't you think?

MBR is in first sector, so you need to skip first sector (512 bytes) of BleskOS image and also leave first sector in USB untouched and rest could be updated just by rewriting it by dd.

BTW it looks like that USB flash that was not working with previous versions of BleskOS is working now. Would you please check also other keys that did not work before, so I may add them to List of working devices?

**Comment by defdefred** [comment]
Yes I will check all my keys :-)
System log output is enough or I need to note the keys model?

**Comment by VendelinSlezak** [comment]
System log is enough, because it provides me Vendor ID and Device ID.

**Comment by defdefred** [comment]
Here is lots of usb logs.
Seems that there is many noname chinese usb flash using ABCD and 1234...
[lotsofusblog.txt](https://github.com/VendelinSlezak/BleskOS/files/13836114/lotsofusblog.txt)

Updating bleskOS without destroying ;y second partition works with:
```
unzip bleskos_2024u3_img.zip 
dd if=bleskos_2024u3.img skip=1 of=/dev/sdb seek=1
```

I tried 16bbp screen too and it work!

**Comment by VendelinSlezak** [comment]
Thank you very much. It will surely take some time to get everything working :) I noticed that two USB keys (according to Vendor ID Verbatim and Espressif Incorporated) did not send USB Inquiry. I made change in code about this, would you please mind to test these two again with Update 4?

**Reply by defdefred** [reply]
Expressif is a virtual usb mass storage fat12 emulated on esp32 microcontroler.
I'm trying to make a mass storage access to internet with the esp32 embedded wifi ...
Another silly thinking ;-).
Could be a mass storage proxy to cloud storage lol.

**Comment by defdefred** [comment]
And verbatim is an external usb mecanical HDD.
[verbatimHDexpressifESP32.txt](https://github.com/VendelinSlezak/BleskOS/files/13851648/verbatimHDexpressifESP32.txt)

I have also new usb device with some usb flash card reader:
[moreusbflashandflashcardadapt.txt](https://github.com/VendelinSlezak/BleskOS/files/13851650/moreusbflashandflashcardadapt.txt)
